Add validation and type checking

pluflou opened the pull request:
This PR addresses #120. Type checking and input validation is added to the following user-facing classes/functions in `mfx.optimize`:
- `Align` and `align.beam`
- `xopt_scans.get_xopt_obj`
- `blop_scans.get_blop_agent`

Notes/questions:
- with the current defaults having no constraints set for the wave8 detectors, Xopt will fail to run in safe mode. This should be resolved once the new detectors are calibrated and the constraints are updated.
- ~the "ip" location is missing in the `align` module, but is listed in `get_xopt_obj`. Is this something that needs to be adjusted?~ discussed on 4/8, issues created
- ~is blop  only setup up with wave8 and no yag on purpose or does this need development?~ discussed on 4/8, needs development

Remaining to-dos in this PR:
- [X] add validation to `blop_scans` module, similarly to `xopt_scans`
- [X] merge #144 first before adjusting/merging this one
- (edit: will do in another PR) add unit tests that cover all expected data types and potential causes of failures (nans, etc)

ZLLentz commented on mfx/optimize/align.py:
```diff
@@ -1,52 +1,96 @@
+
+        # Call the original function with the modified arguments, validated by Pydantic
+        validated_func = validate_call(func, config=ConfigDict(validate_default=True))
+        return validated_func(*new_args, **new_kwargs)
```
I'm a fan of this decorator approach :+1:

ZLLentz commented on mfx/optimize/align.py:
```diff
@@ -1,52 +1,96 @@
-        self.beam_alignment_diagnostics = ["XCS1", "DG1", "DG2"]
-        self.beam_alignment_methods = ["Xopt", "blop"]
-        self.beam_alignment_devices = ["yag", "wave8"]
+    @validate_call
+    def __init__(self, mirror_pitch: list[float] = [-549.0, -546.0]):
```
I understand this is helpful, I'm curious if there were specific incidents that led to us wanting to do more up-front type checking?

pluflou replied:
It was mainly [this](https://stanford.enterprise.slack.com/archives/C06FCE93VAQ/p1741274818766989?thread_ts=1741274818.766989&cid=C06FCE93VAQ) that sparked this work. The idea is to make this "tired-operator-fool-proof".

ZLLentz reviewed (commented):
The diff here is polluted with the unmerged #131 but I'm in agreement
I'll review again if we merge the other one and I can see the diff more clearly

pluflou commented:
> The diff here is polluted with the unmerged #131 but I'm in agreement
> I'll review again if we merge the other one and I can see the diff more clearly

Since you approve the other PR I left those in here. Is there something holding up that merge?

ZLLentz commented:
I'm actually not sure who should have final say on the PRs in mfx, I don't "feel" like I'm the primary maintainer.
If there are no objections I'll merge the other PR on Monday.

pluflou commented:
> I'm actually not sure who should have final say on the PRs in mfx, I don't "feel" like I'm the primary maintainer.
> If there are no objections I'll merge the other PR on Monday.

Oh ok. I wasn't sure if I was supposed to merge after approval or if someone else does it. Sounds good. Thank you!

fredericpoitevin commented:
> > I'm actually not sure who should have final say on the PRs in mfx, I don't "feel" like I'm the primary maintainer.
> > If there are no objections I'll merge the other PR on Monday.
> 
> Oh ok. I wasn't sure if I was supposed to merge after approval or if someone else does it. Sounds good. Thank you!

Sounds like it is time we discuss the SOP, as suggested by Zach. I've started a discussion in #142 for that.

fredericpoitevin commented on mfx/optimize/xopt_scans.py:
```diff
@@ -173,33 +180,68 @@ def get_xopt_obj(
         One of "yag" or "wave8"
     location : str
-        One of "xcs1", "dg1", "dg2", "ip"
+        One of "xcs1", "dg1", "dg2", "ip" # TODO: should IP be added to Diagnostics?
```
Good question. Typically, when we check the beam at the interaction point (ip) with the sample, we considered it "aligned" already and start working on "focusing" it.

In practice, at the moment we can put a YAG at the IP and we routinely focus the beam "by eye" or through a transfocator scan after having put the "right" combination of lenses for the photon energy of the day. The code for that scan should be changed to use lcls_tools (#137 ).
We are also interested in exploring the use of non-invasive IPMs for that task (#138). 

Altogether, I think we should discuss creating a `focus.beam()` utility that would mirror the `align.beam()` code structure, and would replace the existing transfocator scan (**add link to code here**).

fredericpoitevin replied:
see #146

fredericpoitevin commented on mfx/optimize/xopt_scans.py:
```diff
@@ -173,33 +180,68 @@ def get_xopt_obj(
+        Constraint on maximum wave8 xpos for data to be valid
     yag_size_min : float, optional
-        Constraint on minimum yag spot size for data to be valid
+        Constraint on minimum yag spot size (RMS) for data to be valid
```
continuing my comment above, the spot size is something that we check at the ip with the goal of minimizing it.

fredericpoitevin reviewed (approved):
Thanks @pluflou ! Tired operators will owe you one! I'll approve but counting on @ZLLentz to provide more insightful feedbacks while we converge on a SOP (#142)

ZLLentz commented on mfx/optimize/align.py:
```diff
@@ -1,52 +1,96 @@
+        on_diagnostic : str, optional
+            Diagnostic to use for alignment. Options: "XCS1, DG1, DG2". Default is "dg1".
+        with_method : str, optional
+            Method to use for alignment. Options: "blop, Xopt". Default is "xopt".
```
Nitpick: if we're forcing everything lowercase, let's be consistent in the docstring so we don't accidentally have people putting capital strings in when they don't need to.

```suggestion
            Diagnostic to use for alignment. Options: "xcs1, dg1, dg2". Default is "dg1".
        with_method : str, optional
            Method to use for alignment. Options: "blop, xopt". Default is "xopt".
```

ZLLentz commented on mfx/optimize/align.py:
```diff
@@ -1,52 +1,96 @@
+
+def validate_w_lowercase_args(func):
+    """
+   Decorator to make string inputs lowercase before validating.
```
Nitpick: docstring indentation matching
Suggested rewording to make it more clear that this calls a validator

```suggestion
    Decorator to make string inputs lowercase, and then validate.
```

ZLLentz commented on mfx/optimize/blop_scans.py:
```diff
@@ -159,7 +160,7 @@ def get_blop_agent(
         Objective(
             name=df_name,
-            trust_domain=(-1 * wave8_max_value, wave8_max_value),
+            constraint=(-1 * wave8_max_value, wave8_max_value),
```
Question for my understanding after re-reading the blop docs: https://nsls-ii.github.io/blop/objectives.html
- `trust_domain` means "ignore points outside this range" (make them nan)
- `constraint` is not mentioned in the docs but seems to be an alternative to `target`

What motivated this change?

mrakitin replied:
To be addressed via https://github.com/NSLS-II/blop/issues/111.

Maybe @thomaswmorris can confirm the meaning of the `constrain` here?

ZLLentz commented on mfx/optimize/xopt_scans.py:
```diff
@@ -173,33 +180,68 @@ def get_xopt_obj(
+    centroid_x_max : float, optional
+        Constraint on maximum centroid x value for data to be valid
+    centroid_y_min : float, optional
+    Constraint on minimum centroid y value for data to be valid
```
Docs spacing nitpick
```suggestion
        Constraint on minimum centroid y value for data to be valid
```

ZLLentz commented on mfx/optimize/xopt_scans.py:
```diff
@@ -173,33 +180,68 @@ def get_xopt_obj(
+    if device_type == "wave8":
+        if location == "ip":
+            raise ValueError("There is no wave8 at the ip.")
+        if {yag_size_min, yag_size_max, yag_intensity_min, yag_intensity_max} != {None}:
```
Nitpick: This pattern is hard to read because it's indirect.

pluflou replied:
I am changing it to 
```
if any(v is not None for v in (yag_size_min, yag_size_max, yag_intensity_min, yag_intensity_max)):
```

ZLLentz reviewed (commented):
I only have mostly nitpicks here, and mostly in docstrings.
I would have clicked approve if I wasn't confused by the trust_domain vs constraint thing

pluflou committed:
Reapply "merging new changes"

This reverts commit ea65d674f68653f1d860bd954f7ff79d7bab0f0a.

pluflou committed:
Reapply "merging new changes"

This reverts commit 0b6c046ea95bd0d9ab937103c91f2186b902edc5.

pluflou committed:
fix merge conflicts

pluflou committed:
adjust docstrings and yag settings check to improve readability

pluflou commented:
I got a new computer and somehow managed to completely destroy the merge process while setting things up. The commit history is all messed up, but at least the changes make more sense now. Let me know if there's any issues with this PR on your end.

I am working on adding any other validation needed for the stuff Zach added. If we need this merged this week then I can wrap up the blop addition of the yag optimization and leave the unit tests to another PR (probably a good idea anyway).

fredericpoitevin commented:
> I got a new computer and somehow managed to completely destroy the merge process while setting things up. The commit history is all messed up, but at least the changes make more sense now. Let me know if there's any issues with this PR on your end.
> 
> I am working on adding any other validation needed for the stuff Zach added. If we need this merged this week then I can wrap up the blop addition of the yag optimization and leave the unit tests to another PR (probably a good idea anyway).

Yes I think it sounds like a good idea to merge this and create a new issue for the unit tests (important but not urgent).

pluflou committed:
add validation to 2d fns and clean up

pluflou commented:
This PR is ready for final review. I added input validation to the new 2d functions and cleaned up a little.

I propose we leave the following to do in other PRs:
- add tests
- add option to call `xopt_obj_2d_markers` to `align.beam` (maybe makes sense in the refactoring PR)
- add YAG option to the blop scans

fredericpoitevin reviewed (approved):
Thanks @pluflou !
